### PR TITLE
fix(api): validate token permissions in token-create

### LINKED ISSUE #620: Validate token permissions at the token-create API boundary
## Problem

`supabase/functions/token-create/index.ts` casts `body.permissions` to `string[]` without runtime validation. A non-empty string such as `"GP"` passes the required-field check and reaches the array-typed database column, producing a database error and 500 instead of a 400 validation response.

This was surfaced while reviewing formatting-only PR #611. That PR only removes blank lines from this function.

## Suggested direction

Require `permissions` to be an array and validate every element against the supported permission values before insertion. Add negative tests for strings, mixed arrays, empty arrays, and unknown permissions.

### PR DESCRIPTION
**Pull request by DysektAI**
## Summary
- Add `parseTokenPermissions` to the `token-create` Edge Function: `permissions` must be a non-empty array whose elements are all supported permission values (`GP`/`TP`/`WP`), matching the api-gateway `Permission` enum and the frontend `API_PERMISSIONS` map
- Malformed input (a bare string like `"GP"`, mixed arrays, empty arrays, unknown values) now returns `400 permissions must be a non-empty array of: GP, TP, WP` before insertion, instead of a database error/`500` — or inserting permissions the gateway can never honor
- Document the validation in `docs/API.md` (Active Token Cap section)

Closes #620

## Test plan
- [x] New `permissions.test.ts` covers the issue's negative cases — strings, mixed arrays, empty arrays, unknown permissions — plus valid arrays (8 tests)
- [x] `pnpm exec vitest run supabase/functions/token-create/` — 15/15 pass (8 new + 7 existing)
- [x] `deno check token-create/index.ts token-create/permissions.ts` — clean
- [x] Prettier and blank-line lint clean on all touched files

## Summary by CodeRabbit

- **New Features**
  - Token creation now supports validated permission settings using `GP`, `TP`, and `WP`.
  - Permissions must be provided as a non-empty array containing only supported values.

- **Bug Fixes**
  - Invalid permission formats, empty arrays, mixed values, and unsupported permissions are rejected with a clear `400` error before token creation.
  - Invalid request bodies and missing game mode values now receive clearer validation errors.
  - Token creation applies consistent validation and error handling.

- **Documentation**
  - Updated API documentation to explain the required permission format and accepted values.

<details open><summary><h3>Greptile Summary</h3></summary>

The PR validates token permissions before token insertion and synchronizes the public and canonical API documentation.

- Accepts only non-empty arrays containing `GP`, `TP`, or `WP`.
- Returns a clear `400` response for malformed or unsupported permissions.
- Adds focused validation tests and updates `docs/API.md` and `AGENTS.md`.
</details>

<details open><summary><h3>Confidence Score: 5/5</h3></summary>

The PR appears safe to merge.

No blocking failure remains, and the previously reported canonical API guidance issue is fixed at the current head.
</details>

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| AGENTS.md | Adds the canonical token-create permission contract requested by the previous review thread. |
| docs/API.md | Documents the accepted permission shape, supported values, and validation response. |
| supabase/functions/token-create/index.ts | Integrates permission validation before insertion while preserving authentication, rate limiting, and token creation behavior. |
| supabase/functions/token-create/permissions.ts | Defines the permission allowlist and rejects empty, malformed, or unsupported permission inputs. |
| supabase/functions/token-create/permissions.test.ts | Covers valid permission arrays and malformed, empty, mixed, or unsupported inputs. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
  A[Authenticated token-create request] --> B[Apply mutation rate limit]
  B --> C[Parse request body]
  C --> D{Permissions are a non-empty allowlisted array?}
  D -- No --> E[Return 400]
  D -- Yes --> F[Validate game mode and token value]
  F --> G[Insert token]
  G --> H[Return created token]
```
</details>

<sub>Reviews (3): Last reviewed commit: ["fix(api): harden token body parsing and ..."](https://github.com/tarkovtracker-org/tarkovtracker/commit/d26a72baa7adcd9d15f112e18a2d842aa59e9269) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=56686637)</sub>

**Context used:**

- Knowledge Base — [API access and gateway](https://app.greptile.com/dysektai/-/custom-context/knowledge-base/tarkovtracker-org/tarkovtracker/-/docs/api-access-and-gateway.md)

### CODE REVIEW

File: docs/API.md [hunk 1]
```diff
@@ -420,7 +420,7 @@ Polling integrators (TarkovMonitor, tarkov.dev, RatScanner) should poll read end
 
 ### Active Token Cap
 
-Each account may have at most **3 active API tokens**. This is enforced by a database trigger, so token rotation cannot bypass it. The `token-create` Edge Function returns `409` with `error: "Token limit reached (3 active)"` when the cap is reached. Revoke an existing token before creating a new one. Token creation is only allowed through the `token-create` Edge Function (authenticated clients cannot insert into `api_tokens` directly) and is rate-limited to 3 creates per hour per account.
+Each account may have at most **3 active API tokens**. This is enforced by a database trigger, so token rotation cannot bypass it. The `token-create` Edge Function returns `409` with `error: "Token limit reached (3 active)"` when the cap is reached. Revoke an existing token before creating a new one. Token creation is only allowed through the `token-create` Edge Function (authenticated clients cannot insert into `api_tokens` directly) and is rate-limited to 3 creates per hour per account. The `permissions` field must be a non-empty array of `GP`/`TP`/`WP`; any other shape or value is rejected with `400` before insertion.
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply docs/API.md:423]: Addressed in 69808072: AGENTS.md now documents the token-create permissions invariant under State, Data, and APIs — non-empty array of GP/TP/WP matching the api-gateway Permission enum and the frontend API_PERMISSIONS map, rejected with 400 before insertion — per the Maintenance Contract same-PR sync rule.

File: supabase/functions/token-create/permissions.test.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import { describe, expect, it } from 'vitest';
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply supabase/functions/token-create/permissions.test.ts:1]: Not adopting this one. The `*.test.ts` in `__tests__/` naming guidance describes the app-code layout under `app/`; every existing Edge Function test is colocated beside its implementation (`token-create/tokenValue.test.ts`, `_shared/discord.test.ts`, `_shared/stripeBilling.test.ts`, `_shared/stripeTier.test.ts`, plus two `.deno.test.ts` files). Moving only this file would break the established supabase/functions convention without functional benefit.

File: supabase/functions/token-create/index.ts [hunk 3]
```diff
@@ -44,13 +45,20 @@ Deno.serve(async (req) => {
...
+    const permissions = parseTokenPermissions(body.permissions);
     const gameMode = (body.gameMode as string) || '';
     const note = (body.note as string | null) || null;
     let tokenValue = (body.tokenValue as string | undefined) || '';
-    if (!permissions.length || !gameMode) {
+    if (!gameMode) {
       return createErrorResponse('gameMode and permissions are required', 400, req);
     }
+    if (!permissions) {
+      return createErrorResponse(
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply supabase/functions/token-create/index.ts]: Fixed in 69808072 alongside the duplicate Codex finding: AGENTS.md State/Data/APIs now records the enforced token-create permissions contract.

File: supabase/functions/token-create/index.ts [hunk 4]
```diff
@@ -7,90 +8,161 @@ import {
...
+  }
+};
+const parseRequestFields = (body: Record<string, unknown>): TokenRequestFields => ({
+  permissions: parseTokenPermissions(body.permissions),
+  gameMode: (body.gameMode as string) || '',
+  note: (body.note as string | null) || null,
+  tokenValue: (body.tokenValue as string | undefined) || '',
+});
+const gameModePresenceError = (gameMode: string): string | null =>
+  gameMode ? null : 'gameMode and permissions are required';
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply supabase/functions/token-create/index.ts]: Fixed in d26a72ba: the message is now `gameMode is required`. Valid — after this PR split the combined check, invalid permissions carry their own dedicated 400, so naming permissions in the gameMode-presence message became misleading. No test, doc, or client references the old string (verified by repo-wide grep), so the correction is contract-safe.

File: supabase/functions/token-create/index.ts [hunk 5]
```diff
@@ -7,90 +8,161 @@ import {
...
+};
+type MutationGuard =
+  { response: Response } | { user: AuthSuccess['user']; supabase: SupabaseClient };
+const parseJsonBody = async (req: Request): Promise<Record<string, unknown>> => {
+  try {
+    return (await req.json()) as Record<string, unknown>;
+  } catch {
+    return {};
+  }
+};
```
[the remark this answers is not in the document]
Comment (DysektAI) [reply supabase/functions/token-create/index.ts:46]: Fixed in d26a72ba: `parseJsonBody` now returns `{}` for any non-object JSON root (`null`, arrays, primitives) via an `isJsonObject` guard, so these malformed bodies get the required-field `400` instead of an unhandled 500. Kept below the fallow complexity gate by extracting the predicate.
